### PR TITLE
Configuration option to preserve blank lines at start of block

### LINKED ISSUE #2868: Configuration option to preserve blank lines at beginning of block
It would be nice if `rustfmt` had a configuration setting for preserving blank lines at the beginning of blocks.

For example, it would be nice if `rustfmt` converted this program…

```rustfmt
fn main() {
    
    /* blah bleh bluh */
        println!("Hello world");
}
```

…into the following program:

```rustfmt
fn main() {
    
    /* blah bleh bluh */
    println!("Hello world");
}
```

This preservation of space is especially important when the function signature spans several lines, as sometimes it's nice to have a visual break between a complex signature and the body of code.

There once was a configuration option called `remove_blank_lines_at_start_or_end_of_block` that did this, but it was removed as part of #1974 for reasons unapparent.

### PR DESCRIPTION
**Pull request by ayazhafiz**
Adds the option for

- Functions
- Control flow blocks
- Expression statements (i.e. `let x = { ... };`)
- Traits
- Trait Impls
- Mods
- Foreign Mods

If I am missing a kind of item, let me know :)

Closes #2868

### CODE REVIEW

**Comment by ayazhafiz** [comment]
(don't let the 900 line diff scare you; ~800 of it is tests)

**Comment by calebcartwright** [comment]
> (don't let the 900 line diff scare you; ~800 of it is tests)

That made me chuckle, as my eyes did pop at first before I saw that note :eyes:

**Comment by calebcartwright** [comment]
I know the original issue only referenced start of blocks, but I have a suspicion that a decent size chunk of users that would leverage this config option may also be interested in something similar for the end of blocks.

Do you think an equivalent end-of-block config and behavior would be feasible, and if so would it make sense for those to be two distinct config options or easier to use as one? Based on the explanation in the linked issue it sounds like there's scenarios where folks may want to keep the lines at the beginning but still trim at the end so separate would probably be best for more granular control.

**Comment by ayazhafiz** [comment]
That makes sense to me, if people want one they may want the other. I definitely think it's feasible (and probably easily doable, but those are famous last words). However I also think it's worth waiting and seeing if there are any requests for such a configuration option, as it's easier to add an option than remove one (from a stability and "how many people are using this?" point of view). What do you think?

**Comment by ayazhafiz** [comment]
This is totally orthogonal, but it would be interesting to pull GitHub data and see what are the most common rustfmt options are too.

**Comment by calebcartwright** [comment]
> However I also think it's worth waiting and seeing if there are any requests for such a configuration option, as it's easier to add an option than remove one (from a stability and "how many people are using this?" point of view). What do you think?

Agreed, and certainly no need to do anything with end of block lines as part of this PR, I'm betting it'll get requested sooner or later though.

> but it would be interesting to pull GitHub data and see what are the most common rustfmt options are too.

Definitely! Funny you should mention this, as just the other day I was thinking about how beneficial it would be to have quantitative data about config option usage. There's some that we know are used pretty extensively in their respective corners (`indent_style`, `hard_tabs`, `use_small_heuristics`, etc.) but there's plenty of others where we don't really know.

File: src/formatting/items.rs [hunk 1]
```diff
@@ -854,12 +866,25 @@ pub(crate) fn format_impl(
             visitor.block_indent = item_indent;
             visitor.last_pos = lo + BytePos(open_pos as u32);
 
+            let open_nls = &visitor.trim_spaces_after_opening_brace(
+                inner_attributes(&item.attrs)
+                    .first()
+                    .map(|a| a.span.lo())
+                    .or(items.first().map(|i| i.span().lo())),
```
Comment (calebcartwright) [inline src/formatting/items.rs:873]: Bit of a nit but could we use `map_or` here or even `map_or_else` to get the lazy evaluation? 

Something like this perhaps (usual disclaimer about inline GH suggestions not being formatted correctly)
```suggestion
                inner_attributes(&item.attrs).first().map_or_else(
                    || items.first().map(|i| i.span().lo()),
                    |a| Some(a.span.lo()),
                ),
```

File: src/formatting/visitor.rs [hunk 2]
```diff
@@ -181,18 +181,56 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
     }
 
     /// Remove spaces between the opening brace and the first statement or the inner attribute
-    /// of the block.
-    fn trim_spaces_after_opening_brace(
+    /// of the block, fast-forwarding this position of the visitor.
+    /// If `preserve_block_start_blank_lines` is true, the return value contains the newlines that
+    /// should be preserved.
+    pub(crate) fn trim_spaces_after_opening_brace(
```
Comment (calebcartwright) [inline src/formatting/visitor.rs:187]: I kinda feel like we should rename this function to better represent what it's doing now. Whereas before it was indeed trimming things, it's now responsible for determining the newline story between the opening brace and first element within the block. 

What do you think?

File: src/formatting/visitor.rs [hunk 3]
```diff
@@ -181,18 +181,56 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
...
+                        };
+                    if has_multiple_blank_lines {
+                        let mut lines = snippet.lines().map(&str::trim);
+                        lines.next(); // Eat block-opening newline
+                        while let Some("") = lines.next() {
+                            result.push('\n');
+                        }
+                    }
+                } else {
+                    debug!("Failed to preserve blank lines for {:?}", missing_span);
```
Comment (calebcartwright) [inline src/formatting/visitor.rs:230]: Did you encounter this scenario at all during your testing, or is this a defensive/just-in-case?

File: src/formatting/visitor.rs [hunk 4]
```diff
@@ -945,6 +989,14 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
             } else {
                 self.last_pos = mod_lo;
                 self.block_indent = self.block_indent.block_indent(self.config);
+
+                let first_non_ws = inner_attributes(attrs)
+                    .first()
+                    .map(|attr| attr.span.lo())
+                    .or(m.items.first().map(|s| s.span().lo()));
```
Comment (calebcartwright) [inline src/formatting/visitor.rs:996]: Same thought as above with a `map_or` or `map_or_else`

File: src/formatting/visitor.rs [hunk 5]
```diff
@@ -225,7 +264,12 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
         self.last_pos = self.last_pos + brace_compensation;
         self.block_indent = self.block_indent.block_indent(self.config);
         self.push_str("{");
-        self.trim_spaces_after_opening_brace(b, inner_attrs);
+
+        let first_non_ws = inner_attrs
+            .and_then(|attrs| inner_attributes(attrs).first().map(|attr| attr.span.lo()))
+            .or(b.stmts.first().map(|s| s.span().lo()));
```
Comment (calebcartwright) [inline src/formatting/visitor.rs:270]: Same thought as above if `or_else` could be used

File: src/formatting/items.rs (hunk 1 above)
Comment (ayazhafiz) [reply src/formatting/items.rs:873]: Done

File: src/formatting/visitor.rs (hunk 3 above)
Comment (ayazhafiz) [reply src/formatting/visitor.rs:230]: No, just adding as a foresight

File: src/formatting/visitor.rs (hunk 2 above)
Comment (ayazhafiz) [reply src/formatting/visitor.rs]: Good point, renamed to `advance_to_first_block_item`

**Comment by calebcartwright** [comment]
Thanks for the updates! I'll try to get back to this soon, though have also requested a review from @topecongiro for the reasons mentioned earlier in the thread. 

One of the things that I can't quite shake is a feeling that there's a little too much going on with `advance_to_first_block_item` and/or some of the internal logic is leaking out a bit that has to be conditionally interpreted by consumers of the function.

For example, I can see 3 different branches internal to the function that can all result in function returning the exact same thing, an empty string, and the length of the returned string (or more precisely whether or not it's empty) is significant for at least some consumers. 

I don't have any specific suggestions/modifications just yet (although maybe wrapping the return string in an Option could help?) and I wouldn't say this is a blocking concern for me; just wanted to share my thinking. Properly (and idempotently) maintaining line spacing is certainly a nontrivial task and you've done a great job here!

**Comment by ayazhafiz** [comment]
> a little too much going on with `advance_to_first_block_item`

I tried to condense the logic a bit, lmk what you think

**Comment by ayazhafiz** [comment]
https://github.com/rust-lang/rustfmt/pull/4295/files broke this, so there is probably a better way to handle this using the code there...

**Comment by calebcartwright** [comment]
Thanks for the changes @ayazhafiz! I know it was fairly minor but I definitely appreciate the updated  `advance_to_first_block_item` :+1:

**Pull request merged**